File: src/test/java/org/micoli/php/MessengerServiceTest.java
package org.micoli.php;

import com.intellij.codeInsight.daemon.GutterMark;
import com.intellij.testFramework.fixtures.BasePlatformTestCase;
import com.jetbrains.php.lang.psi.elements.Method;
import com.jetbrains.php.lang.psi.elements.MethodReference;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.PhpNamedElement;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import org.micoli.php.configuration.ConfigurationException;
import org.micoli.php.configuration.ConfigurationFactory;
import org.micoli.php.configuration.NoConfigurationFileException;
import org.micoli.php.service.PhpUtil;
import org.micoli.php.symfony.messenger.configuration.SymfonyMessengerConfiguration;
import org.micoli.php.symfony.messenger.service.MessengerService;

public class MessengerServiceTest extends BasePlatformTestCase {

    @Override
    protected String getTestDataPath() {
        return "src/test/resources/testData";
    }

    public void testItDetectMessageBasedOnPatternClass() {
        myFixture.configureByFile("/src/Core/Event/ArticleCreatedEvent.php");
        SymfonyMessengerConfiguration symfonyMessengerConfiguration = new SymfonyMessengerConfiguration();
        symfonyMessengerConfiguration.messageClassNamePatterns = ".*(edEvent|Command)$";
        MessengerService messengerService = MessengerService.getInstance(getProject());
        messengerService.loadConfiguration(myFixture.getProject(), symfonyMessengerConfiguration);
        PhpClass phpClass = PhpUtil.getPhpClassByFQN(getProject(), "App\\Core\\Event\\ArticleCreatedEvent");
        assertNotNull(phpClass);
        assertTrue(messengerService.isMessageClass(phpClass));
    }

    public void testItDoesNotDetectMessageBasedOnPatternIfPatternIsWrong() {
        myFixture.configureByFile("/src/Core/Event/ArticleCreatedEvent.php");
        SymfonyMessengerConfiguration symfonyMessengerConfiguration = new SymfonyMessengerConfiguration();
        symfonyMessengerConfiguration.messageClassNamePatterns = ".*(Command)$";
        MessengerService messengerService = MessengerService.getInstance(getProject());
        messengerService.loadConfiguration(myFixture.getProject(), symfonyMessengerConfiguration);
        PhpClass phpClass = PhpUtil.getPhpClassByFQN(getProject(), "App\\Core\\Event\\ArticleCreatedEvent");
        assertNotNull(phpClass);
        assertFalse(messengerService.isMessageClass(phpClass));
    }

    public void testItDetectMessageBasedOnInterface() {
        myFixture.configureByFiles(
                "/src/Core/Event/ArticleCreatedEvent.php",
                "/src/Infrastructure/Bus/Message/Event/AsyncEventInterface.php",
                "/src/Infrastructure/Bus/Message/Event/EventInterface.php",
                "/src/Infrastructure/Bus/Message/MessageInterface.php");
        SymfonyMessengerConfiguration symfonyMessengerConfiguration = new SymfonyMessengerConfiguration();
        symfonyMessengerConfiguration.messageInterfaces =
                new String[] {"App\\Infrastructure\\Bus\\Message\\MessageInterface"};
        MessengerService messengerService = MessengerService.getInstance(getProject());
        messengerService.loadConfiguration(myFixture.getProject(), symfonyMessengerConfiguration);
        PhpClass phpClass = PhpUtil.getPhpClassByFQN(getProject(), "App\\Core\\Event\\ArticleCreatedEvent");
        assertNotNull(phpClass);
        assertTrue(messengerService.isMessageClass(phpClass));
    }

    public void testItCanFindHandlersByMessage() {
        myFixture.copyDirectoryToProject("src", "src");
        MessengerService messengerService = loadPluginConfiguration(getTestDataPath());
        Collection<Method> handledMessages =
                messengerService.findHandlersByMessageName("App\\Core\\Event\\ArticleCreatedEvent");
        assertContainsElements(
                handledMessages.stream().map(PhpNamedElement::getFQN).toList(),
                "\\App\\Core\\EventListener\\OnArticleCreated.__invoke");
    }

    public void testItCanFindLineMarkersForMessageHandler() {
        myFixture.copyDirectoryToProject("src", "src");
        myFixture.configureByFiles("src/Core/EventListener/OnFeedCreated.php");
        loadPluginConfiguration(getTestDataPath());
        List<GutterMark> lineMarkers = myFixture.findAllGutters();
        assertNotEmpty(lineMarkers);

        List<GutterMark> specificMarkers = lineMarkers.stream()
                .filter(it -> {
                    String tooltipText = it.getTooltipText();
                    if (tooltipText == null) {
                        return false;
                    }
                    return tooltipText.contains("Search for usages")
                            || tooltipText.contains("Navigate to message handlers");
                })
                .toList();
        assertEquals(2, specificMarkers.size());
    }

    public void testItCanFindDispatchCallsForMessageClass() {
        myFixture.copyDirectoryToProject("src", "src");
        MessengerService messengerService = loadPluginConfiguration(getTestDataPath());
        Collection<MethodReference> callsWithoutRootNamespace =
                messengerService.findDispatchCallsForMessage("App\\Core\\Event\\ArticleCreatedEvent");
        Collection<MethodReference> callsWithRootNamespace =
                messengerService.findDispatchCallsForMessage("\\App\\Core\\Event\\ArticleCreatedEvent");

        assertEquals(1, callsWithoutRootNamespace.size());
        assertEquals(callsWithRootNamespace.size(), callsWithoutRootNamespace.size());
    }

    private MessengerService loadPluginConfiguration(String path) {
        SymfonyMessengerConfiguration symfonyMessengerConfiguration = null;
        try {
            symfonyMessengerConfiguration = Objects.requireNonNull(ConfigurationFactory.loadConfiguration(path, 0L))
                    .configuration
                    .symfonyMessenger;
        } catch (ConfigurationException | NoConfigurationFileException e) {
            throw new RuntimeException(e);
        }
        MessengerService messengerService = MessengerService.getInstance(getProject());
        messengerService.loadConfiguration(myFixture.getProject(), symfonyMessengerConfiguration);

        return messengerService;
    }
}

File: src/test/kotlin/org/micoli/php/MessengerServiceTest.kt
package org.micoli.php

import com.intellij.testFramework.fixtures.BasePlatformTestCase
import org.assertj.core.api.Assertions.*
import org.micoli.php.configuration.ConfigurationException
import org.micoli.php.configuration.ConfigurationFactory
import org.micoli.php.configuration.exceptions.NoConfigurationFileException
import org.micoli.php.service.intellij.psi.PhpUtil
import org.micoli.php.symfony.messenger.configuration.SymfonyMessengerConfiguration
import org.micoli.php.symfony.messenger.service.MessengerService

class MessengerServiceTest : BasePlatformTestCase() {
    override fun getTestDataPath(): String = "src/test/resources/symfony-demo"

    fun testItDetectMessageBasedOnPatternClass() {
        myFixture.configureByFile("/src/UseCase/ArticleViewed/Event.php")
        val symfonyMessengerConfiguration = SymfonyMessengerConfiguration()
        symfonyMessengerConfiguration.messageClassNamePatterns = ".*(Event|Command)$"
        val messengerService = MessengerService.getInstance(project)
        messengerService.loadConfiguration(symfonyMessengerConfiguration)
        val phpClass = PhpUtil.getPhpClassByFQN(project, "App\\UseCase\\ArticleViewed\\Event")
        assertThat(phpClass).isNotNull
        assertThat(messengerService.isMessageClass(phpClass!!)).isTrue
    }

    fun testItDoesNotDetectMessageBasedOnPatternIfPatternIsWrong() {
        myFixture.configureByFile("/src/UseCase/ArticleViewed/Event.php")
        val symfonyMessengerConfiguration = SymfonyMessengerConfiguration()
        symfonyMessengerConfiguration.messageClassNamePatterns = ".*(Command)$"
        val messengerService = MessengerService.getInstance(project)
        messengerService.loadConfiguration(symfonyMessengerConfiguration)
        val phpClass = PhpUtil.getPhpClassByFQN(project, "App\\UseCase\\ArticleViewed\\Event")
        assertNotNull(phpClass)
        assertFalse(messengerService.isMessageClass(phpClass!!))
    }

    fun testItDetectMessageBasedOnInterface() {
        myFixture.configureByFiles(
            "/src/UseCase/ArticleViewed/Event.php",
            "/src/Infrastructure/Bus/Message/Event/DomainEvent.php",
            "/src/Infrastructure/Bus/Message/Command/DomainCommand.php",
            "/src/Infrastructure/Bus/Message/Command/SyncDomainCommandResult.php",
            "/src/Infrastructure/Bus/Message/Query/SyncDomainQueryResult.php",
            "/src/Infrastructure/Bus/Message/Command/SyncDomainCommand.php",
            "/src/Infrastructure/Bus/Message/Event/SyncDomainEvent.php",
            "/src/Infrastructure/Bus/Message/DomainMessage.php",
            "/src/Infrastructure/Bus/Message/Query/SyncDomainQuery.php",
            "/src/Infrastructure/Bus/Message/Query/DomainQuery.php",
            "/src/Infrastructure/Bus/Message/Command/SyncDomainCommandWithResult.php",
            "/src/Infrastructure/Bus/Message/Event/AsyncDomainEvent.php",
            "/src/Infrastructure/Bus/Message/Command/AsyncDomainCommand.php",
        )
        val symfonyMessengerConfiguration = SymfonyMessengerConfiguration()
        symfonyMessengerConfiguration.messageInterfaces =
            arrayOf("App\\Infrastructure\\Bus\\Message\\DomainMessage")
        val messengerService = MessengerService.getInstance(project)
        messengerService.loadConfiguration(symfonyMessengerConfiguration)
        val phpClass = PhpUtil.getPhpClassByFQN(project, "App\\UseCase\\ArticleViewed\\Event")
        assertThat(phpClass).isNotNull
        assertThat(messengerService.isMessageClass(phpClass!!)).isTrue
    }

    fun testItCanFindHandlersByMessage() {
        myFixture.copyDirectoryToProject("src", "src")
        val messengerService = loadPluginConfiguration(testDataPath)
        val handledMessages =
            messengerService.findHandlersByMessageName("App\\UseCase\\ArticleViewed\\Event")
        assertThat(handledMessages.stream().map { it!!.fqn }.toList())
            .contains("\\App\\UseCase\\ArticleViewed\\Handler.__invoke")
    }

    fun testItCanFindLineMarkersForMessageHandler() {
        myFixture.copyDirectoryToProject("src", "src")
        myFixture.configureByFiles("src/UseCase/ArticleViewed/Handler.php")
        loadPluginConfiguration(testDataPath)
        val lineMarkers = myFixture.findAllGutters()
        assertThat(lineMarkers).isNotEmpty

        val specificMarkers =
            lineMarkers
                .stream()
                .filter {
                    val tooltipText = it!!.tooltipText ?: return@filter false
                    tooltipText.contains("Search for usages") ||
                        tooltipText.contains("Navigate to message handlers")
                }
                .toList()
        assertThat(specificMarkers.size).isEqualTo(1)
    }

    fun testItCanFindDispatchCallsForMessageClass() {
        myFixture.copyDirectoryToProject("src", "src")
        val messengerService = loadPluginConfiguration(testDataPath)
        val callsWithoutRootNamespace =
            messengerService.findDispatchCallsForMessage("App\\UseCase\\ArticleViewed\\Event")
        val callsWithRootNamespace =
            messengerService.findDispatchCallsForMessage("\\App\\UseCase\\ArticleViewed\\Event")

        assertThat(callsWithoutRootNamespace.size).isEqualTo(2)
        assertThat(callsWithoutRootNamespace.size).isEqualTo(callsWithRootNamespace.size)
    }

    private fun loadPluginConfiguration(path: String?): MessengerService {
        try {
            val symfonyMessengerConfiguration =
                ConfigurationFactory()
                    .loadConfiguration(path, 0L, true)
                    ?.configuration
                    ?.symfonyMessenger
            val messengerService = MessengerService.getInstance(project)
            messengerService.loadConfiguration(symfonyMessengerConfiguration)

            return messengerService
        } catch (e: ConfigurationException) {
            throw RuntimeException(e)
        } catch (e: NoConfigurationFileException) {
            throw RuntimeException(e)
        }
    }
}
